### PR TITLE
feat: market/chain: fip 0027 integration

### PR DESCRIPTION
**Pull request by ZenGround0**
## Related Issues

FIP-0027 integration

## Proposed Changes

FIP-0027 integration from specs-actors
* New deal proposal type
* New migration

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Comment by ZenGround0** [comment]
CI failures now match the base branch

**Comment by arajasek** [comment]
Okay, this looks good too!

File: markets/storageadapter/dealpublisher.go [hunk 1]
```diff
@@ -60,7 +58,7 @@ type DealPublisher struct {
 	publishSpec           *api.MessageSendSpec
 
 	lk                      sync.Mutex
-	pending                 []*pendingDeal
+	pending                 []*pendingDeal ``
```
Comment (magik6k) [inline markets/storageadapter/dealpublisher.go:61]: Why have
```
``
```
?

**Review by magik6k — APPROVED** [review]

File: markets/storageadapter/client.go [hunk 2]
```diff
@@ -368,7 +368,7 @@ func (c *ClientNodeAdapter) OnDealExpiredOrSlashed(ctx context.Context, dealID a
 	return nil
 }
 
-func (c *ClientNodeAdapter) SignProposal(ctx context.Context, signer address.Address, proposal market0.DealProposal) (*marketactor.ClientDealProposal, error) {
```
Comment (magik6k) [inline markets/storageadapter/client.go:371]: I'm assuming that at least for valid string labels cbor serialization stays the same, so we're not worrying about invalid signatures with current actors?

Comment (ZenGround0) [reply markets/storageadapter/client.go:371]: Correct.  And if I understand you right that's the entire point of all this union type hoop-jumping.

File: markets/storageadapter/dealpublisher.go (hunk 1 above)
Comment (ZenGround0) [reply markets/storageadapter/dealpublisher.go]: Weird. Removed

File: go.mod [hunk 3]
```diff
@@ -38,24 +38,24 @@ require (
...
-	github.com/filecoin-project/go-fil-markets v1.20.1
+	github.com/filecoin-project/go-fil-markets v1.20.2
 	github.com/filecoin-project/go-jsonrpc v0.1.5
-	github.com/filecoin-project/go-legs v0.3.7
+	github.com/filecoin-project/go-legs v0.3.10
 	github.com/filecoin-project/go-padreader v0.0.1
 	github.com/filecoin-project/go-paramfetch v0.0.4
 	github.com/filecoin-project/go-state-types v0.1.3
-	github.com/filecoin-project/go-statemachine v1.0.1
+	github.com/filecoin-project/go-statemachine v1.0.2-0.20220322104818-27f8fbb86dfd
```
Comment (magik6k) [inline go.mod:47]: Will need a tag before getting into master

**Review by magik6k — APPROVED** [review]

**Pull request merged**